### PR TITLE
Scala.js implementation & Windows build

### LINKED ISSUE #75: Windows Support
We need to offer Windows support for the CLI.

### PR DESCRIPTION
**Pull request by longshorej**
This uses [sbt-crossproject](https://github.com/portable-scala/sbt-crossproject) to provide JS and Native implementations of the CLI with shims for each defining the platform specific code.

It also adds a build file to produce .zip artifacts for Windows.

CLI had to be refactored to be based around async I/O, given JS's event loop. Hence, many of the types are `Future` now.
  
Fixes #75

### CODE REVIEW

**Comment by longshorej** [comment]
I'll pick this up tomorrow. Native appears to be working, JS is quite close, see e.g. the help output below. Missing for JS currently is process execution and http requests, otherwise all other shims have been implemented.

```sbt
> cliJS/run
[info] Compiling 1 Scala source to /home/longshorej/work/lightbend/reactive-cli/cli/js/target/scala-2.11/classes...
[info] Fast optimizing /home/longshorej/work/lightbend/reactive-cli/cli/js/target/scala-2.11/reactive-cli-fastopt.js
[info] Running com.lightbend.rp.reactivecli.Main
reactive-cli 0.6.3-SNAPSHOT
Usage: reactive-cli [version|generate-kubernetes-deployment] [options] <args>...

  --cainfo <value>         Path to the CA certs for TLS validation purposes
  -l, --loglevel <value>   Sets the log level. Available: error, warn, info, debug, trace
  --help                   Print this help text
Command: version
Outputs the program's version
Command: generate-kubernetes-deployment [options] docker-image
Generate Kubernetes resource files for kubectl
  docker-image             Docker image to be deployed. Format: [<registry host>/][<repo>/]image[:tag]
  --deployment-type <value>
                           Sets the deployment type. Default: canary; Available: canary, blue-green, rolling
  --env <value>            Sets an environment variable. Format: NAME=value
  --external-service <value>
                           Declares an external service address. Format: NAME=value
  --generate-all           Generate all resource types. This is the default and overrides other generate flags
  --generate-ingress       Generate Ingress resources. When provided, Ingress and Pod Controller resource types are not generated unless explicitly requested
  --generate-namespaces    Generate Namespace resources. Does not affect the generation of other resources
  --generate-pod-controllers
                           Generate PodController resources. When provided, Ingress and Pod Controller resource types are not generated unless explicitly requested
  --generate-services      Generate Service resources. When provided, Ingress and Pod Controller resource types are not generated unless explicitly requested
  --ingress-annotation <value>
                           Adds an annotation to the generated Ingress resources. Format: NAME=value
  --ingress-api-version <value>
                           Sets the Ingress API version. Default: extensions/v1beta1
  --ingress-path-suffix <value>
                           Appends the expression specified to the paths of the generated Ingress resources
  --namespace <value>      Resources will be generated with the supplied namespace
  -o, --output <value>     Specify the output directory for the generated resources
  --pod-controller-api-version <value>
                           Sets the Pod Controller (e.g. Deployment) API version. Default: apps/v1beta2
  --pod-controller-image-pull-policy <value>
                           Sets the Docker image pull policy for Pod Controller resources. Supported: Never, IfNotPresent, Always
  --pod-controller-replicas <value>
                           Sets the number of replicas for the Pod Controller resources
  --registry-disable-https
                           Disables HTTPS when accessing docker registry
  --registry-disable-tls-validation
                           Disables TLS cert validation when accessing docker registry through HTTPS
  --registry-username <value>
                           Specify username to access docker registry. Password must be specified also.
  --registry-password <value>
                           Specify password to access docker registry. Username must be specified also.
  --service-api-version <value>
                           Sets the Service API version. Default: v1
  --service-cluster-ip <value>
                           Sets the cluster IP for Service resources
  --version <value>        Uses specified version tag for generated resources instead of version in the docker image
  --transform-ingress <value>
                           A jq expression that will be applied to Ingress resources. jq must be installed
  --transform-namespaces <value>
                           A jq expression that will be applied to Namespace resources. jq must be installed
  --transform-pod-controllers <value>
                           A jq expression that will be applied to Pod Controller resources. jq must be installed
  --transform-services <value>
                           A jq expression that will be applied to Service resources. jq must be installed
[success] Total time: 3 s, completed Jan 9, 2018 6:03:37 PM
```

**Comment by longmuir** [comment]
Yes, no particular customer at the moment - so no major urgency, but I'm sure it'll come up. Minikube and kubectl both support Windows as well. The ConductR sandbox's lack of Windows support became an issue at IMS, and I believe a couple of other clients as well.

**Comment by longshorej** [comment]
Still WIP, but getting closer:

```
~/work/lightbend/reactive-cli/cli/js/target/scala-2.11#windows $ node ../rp.js generate-kubernetes-deployment akka-cluster-tooling-example/akka-cluster-tooling-example:0.1.0  --pod-controller-replicas 3 | md5sum
9cb4d02a1787ee2013867e6e9a64f2ef  -
-> 0

~/work/lightbend/reactive-cli/cli/js/target/scala-2.11#windows $ rp generate-kubernetes-deployment akka-cluster-tooling-example/akka-cluster-tooling-example:0.1.0  --pod-controller-replicas 3 | md5sum
9cb4d02a1787ee2013867e6e9a64f2ef  -
-> 0
```

**Comment by longshorej** [comment]
I think this is fully working now (just pushed JS process support), but I have some refactoring and tests to fix up.

**Comment by longshorej** [comment]
Okay - this is ready for review. Probably best to checkout the PR locally, try building and running it, etc.

File: build-windows.ps1 [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+nexe -o target\win\rp.exe cli\js\target\rp.js
+
+$nexe_success = $?
+
+if (-not $nexe_success) {
+  throw "nexe failed"
+}
+
+Compress-Archive -Path .\target\win\rp.exe -DestinationPath ".\target\win\reactive-cli-$version-Windows.zip"
```
Comment (longshorej) [inline build-windows.ps1]: Is there a naming convention for Windows zip files? I know we're following one for the macOS/Homebrew stuff. Maybe this should be `reactive-cli-<version>-windows-amd64.zip`?

Comment (longshorej) [reply build-windows.ps1]: Also, wouldn't be surprised if I'm doing something silly in this file, being the first PowerShell script I've written.

Comment (mitkus) [reply build-windows.ps1]: Your PowerShell looks fine to me. And no, there are no conventions for naming archives.

I was running this on Windows 7, which comes with PowerShell 2.0 and it doesn't have Compress-Archive builtin so the script failed. I think it's okay though, we can make releases on a newer OS. Most importantly, rp.exe runs fine on the same Win7.

**Pull request merged**